internal/gps/source.go

```diff
@@ -552,51 +542,127 @@ func (sg *sourceGateway) createSingleSourceCache() singleSourceCache {
 	return newMemoryCache()
 }
 
-func (sg *sourceGateway) require(ctx context.Context, wanted sourceState) (errState sourceState, err error) {
+// setUpSource tries to set src from choices and returns the initial sourceState,
+// otherwise it returns a slice of all errors. If force is true, then each
+// maybeSource may make a second attempt after removing the cache directory. testFn
+// is an optional check which may access src and return additional sourceState.
+func (sg *sourceGateway) setUpSource(ctx context.Context, force bool, testFn func(context.Context) (sourceState, error)) (sourceState, errorSlice) {
+	var errs errorSlice
+	for _, mb := range sg.maybe {
+		var err error
+		sg.src, err = mb.try(ctx, sg.cachedir, force, sg.suprvsr)
+		if err == nil {
+			if testFn == nil {
+				return 0, nil
+			}
+			addlState, err := testFn(ctx)
+			if err == nil {
+				return addlState, nil
+			}
+		}
+		errs = append(errs, err)
+	}
+	return 0, errs
+}
+
+// sourceExistsUpstream verifies that the source exists upstream and returns any
+// additional sourceState, or an error.
+func (sg *sourceGateway) sourceExistsUpstream(ctx context.Context) (addlState sourceState, err error) {
+	if sg.src.existsCallsListVersions() {
+		as, err := sg.loadLatestVersionList(ctx)
+		if err != nil {
+			return 0, err
+		}
+		addlState |= as
+	} else {
+		if err := sg.suprvsr.do(ctx, sg.src.sourceType(), ctSourcePing, func(ctx context.Context) error {
+			if !sg.src.existsUpstream(ctx) {
+				return errors.Errorf("source does not exist upstream: %s: %s", sg.src.sourceType(), sg.src.upstreamURL())
+			}
+			return nil
+		}); err != nil {
+			return 0, err
+		}
+	}
+	//TODO(jmank88) broadcast sg.src.upstreamURL() changes here
```

Comment:
Perhaps now is the time to address this, but I don't have a simple solution.  Remapping the URLs would catch many cases, but it seems like there is always a worst case scenario where two of (or sets of) sources/gateways/caches have to be merged after both being in use.  However, I may be considering a more general case than is necessary.  This partially depends on the possible URLs available from `maybeSource`s, and whether a `source`'s `upstreamURL` is always from this set (meaning we could map the possibilities ahead of time).

Reply:
i think our best solution is likely to be in changing the scopes of responsibility, here. something like the following: we create some new indirection layer that basically acts like a persistent `sourceGateway` factory: it takes a `maybeSources` and occupies _all_ the corresponding URL slots in the `sourceCoordinator`'s URL lookup map, and spits out a `sourceGateway` on request.

now, the `sourceCoordinator` is only responsible for mapping to families of related URLs, and the `sourceGatewayFactory` (or whatever) is doing the final layer of mapping work. it's likely that we'll need to rework some of `sourceCoordinator.getSourceGatewayFor()` to accommodate this, possibly including the way that `sourceCoordinator.srcs` and `sourceCoordinator.nameToURL` actually work.

that expands the scope of this PR significantly, though, and my gut currently says we'll be able to defer it until a later PR. for now, we'll probably be able to get away with just having a hard failure here where we might otherwise be able to recover by going back to other, possible upstream URLs.

Reply:
> for now, we'll probably be able to get away with just having a hard failure here where we might otherwise be able to recover by going back to other, possible upstream URLs.

FWIW I've been running with a `panic` on any url change and it hasn't happened yet (mostly ran tests and cockroachdb).  A regular error is probably sufficient though.

Reply:
> it takes a maybeSources and occupies all the corresponding URL slots in the sourceCoordinator's URL lookup map, and spits out a sourceGateway on request.

Can we be certain about all of the possible URLs ahead of time?  e.g. gopkgin code currently returns one url from `maybeSource`, but a different `aliasURL` from `gopkginSource`.  That one would be trivial to change, and the others look pretty standard and static.  For some reason I had it in my head that it would be possible to ping upstream and be forwarded to another remote or something.  If that is not the case, then we may be able to get away with the existing data structures (with more mapping) and with some mechanism for merging additional `maybeSource`s into a gateway.

Reply:
> FWIW I've been running with a panic on any url change and it hasn't happened yet (mostly ran tests and cockroachdb).

I take that back. Now I'm seeing test failures on CI like: `upstream source url changed from "ssh://git@github.com/sdboyer/gpkt" to "https://github.com/sdboyer/gpkt"` which seems like something we need to handle now.

Reply:
This has been fixed.